File: app/src/main/java/com/example/rssreader/screens/SettingsFragment.java
package com.example.rssreader.screens;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Patterns;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.example.rssreader.MainActivity;
import com.example.rssreader.R;
import com.example.rssreader.models.LinkItem;

import java.util.ArrayList;
import java.util.Objects;


public class SettingsFragment extends PreferenceFragmentCompat {

    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        // Load the Preferences from the XML file
        addPreferencesFromResource(R.xml.root_preferences);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Preference deleteEntry = findPreference(getString(R.string.delete_entry));
        Preference addEntry = findPreference(getString(R.string.add_entry));
        Preference defaults = findPreference(getString(R.string.preferences_edittext_default));

        assert defaults != null;
        assert deleteEntry != null;
        assert addEntry != null;

        defaults.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                AlertDialog.Builder dialog = new AlertDialog.Builder(getContext());
                dialog.setTitle("Reset links to default links?");

                dialog.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        ((MainActivity) Objects.requireNonNull(getActivity())).deleteListFromPreferences();
                        ((MainActivity) getActivity()).setListToDefaults();
                        dialog.dismiss();
                    }
                });

                dialog.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });
                dialog.show();
                return true;
            }
        });

        addEntry.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                //Make new Dialog
                AlertDialog.Builder dialog = new AlertDialog.Builder(getContext());
                dialog.setTitle("Set target Title & link");
                final Context context = getContext();
                LinearLayout layout = new LinearLayout(context);
                layout.setOrientation(LinearLayout.VERTICAL);

                final TextView titleLabel = new TextView(context);
                titleLabel.setText(getString(R.string.preferences_edittext_title));
                layout.addView(titleLabel);

                final EditText titleBox = new EditText(context);
                titleBox.setHint("Title");
                layout.addView(titleBox);

                final EditText linkBox = new EditText(context);
                linkBox.setHint("Link");
                layout.addView(linkBox);
                dialog.setView(layout);

                dialog.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });

                dialog.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String link = linkBox.getText().toString();
                        if (Patterns.WEB_URL.matcher(link.toLowerCase()).matches()) {
                            MainActivity.urls.add(new LinkItem(titleBox.getText().toString(), link));
                            ((MainActivity) Objects.requireNonNull(getActivity())).saveListInPreferences();
                            dialog.dismiss();
                        } else {
                            Toast.makeText(context, "Link format invalid!", Toast.LENGTH_LONG).show();
                        }
                    }
                });

                dialog.show();
                return true;
            }
        });

        deleteEntry.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                AlertDialog.Builder builderSingle = new AlertDialog.Builder(getActivity());
                builderSingle.setIcon(R.drawable.ic_menu_gallery);
                builderSingle.setTitle("Select One Name:-");

                ArrayList<String> newArray = new ArrayList<>();
                for (LinkItem i :
                     MainActivity.urls) {
                    newArray.add(i.Title);
                }

                final ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(Objects.requireNonNull(getActivity()), R.layout.settings_list_item, newArray);

                builderSingle.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });

                builderSingle.setAdapter(arrayAdapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        final int index = which;
                        String strName = arrayAdapter.getItem(which);
                        AlertDialog.Builder builderInner = new AlertDialog.Builder(getActivity());
                        builderInner.setMessage(strName != null ? strName : "error");
                        builderInner.setTitle("Your Selected Item to delete is");
                        builderInner.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                MainActivity.urls.remove(index);
                                ((MainActivity) Objects.requireNonNull(getActivity())).saveListInPreferences();
                                arrayAdapter.notifyDataSetChanged();
                                dialog.dismiss();
                            }
                        });
                        builderInner.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                            }
                        });
                        builderInner.show();
                    }
                });
                builderSingle.show();
                return true;
            }
        });
    }

    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        return super.onPreferenceTreeClick(preference);
    }


}

File: app/src/main/java/com/example/rssreader/screens/SettingsFragment.kt
package com.example.rssreader.screens

import android.app.AlertDialog
import android.os.Bundle
import android.util.Patterns
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.FragmentActivity
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat

import com.example.rssreader.MainActivity
import com.example.rssreader.R
import com.example.rssreader.adapters.LinkItemAdapter
import com.example.rssreader.models.LinkItem

import java.util.Objects


class SettingsFragment : PreferenceFragmentCompat() {

    override fun onCreatePreferences(bundle: Bundle?, s: String?) {
        // Load the Preferences from the XML file
        addPreferencesFromResource(R.xml.root_preferences)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val deleteEntry = findPreference<Preference>(getString(R.string.delete_entry))
        val addEntry = findPreference<Preference>(getString(R.string.add_entry))
        val defaults = findPreference<Preference>(getString(R.string.preferences_edittext_default))!!

        assert(deleteEntry != null)
        assert(addEntry != null)

        defaults.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val dialog = AlertDialog.Builder(context)
            dialog.setTitle("Reset links to default links?")

            dialog.setPositiveButton("Ok") { dialog1, _ ->
                (Objects.requireNonNull<FragmentActivity>(activity) as MainActivity).deleteListFromPreferences()
                dialog1.dismiss()
            }

            dialog.setNegativeButton("cancel") { dialog1, _ -> dialog1.dismiss() }
            dialog.show()
            true
        }

        addEntry!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            //Make new Dialog
            val dialog = AlertDialog.Builder(context)
            dialog.setTitle("Set target Title & link")
            val context = context
            val layout = LinearLayout(context)
            layout.orientation = LinearLayout.VERTICAL

            val titleLabel = TextView(context)
            titleLabel.text = getString(R.string.preferences_edittext_title)
            layout.addView(titleLabel)

            val titleBox = EditText(context)
            titleBox.hint = "Title"
            layout.addView(titleBox) // Notice this is an add method

            val linkBox = EditText(context)
            linkBox.hint = "Link"
            layout.addView(linkBox) // Another add method
            dialog.setView(layout) // Again this is a set method, not add

            dialog.setNegativeButton("cancel") { dialog1, _ -> dialog1.dismiss() }

            dialog.setPositiveButton("Ok") { dialog1, _ ->
                val link = linkBox.text.toString()
                if (Patterns.WEB_URL.matcher(link.toLowerCase()).matches()) {
                    MainActivity.urls!!.add(LinkItem(titleBox.text.toString(), link))
                    (Objects.requireNonNull<FragmentActivity>(activity) as MainActivity).saveListInPreferences()
                    dialog1.dismiss()
                } else {
                    Toast.makeText(context, "Link format invalid!", Toast.LENGTH_LONG).show()
                }
            }

            dialog.show()
            true
        }

        deleteEntry!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val builderSingle = AlertDialog.Builder(activity)
            builderSingle.setIcon(R.drawable.ic_launcher_foreground)
            builderSingle.setTitle("Select One Name:-")

            val arrayAdapter = MainActivity.urls?.let { it1 -> LinkItemAdapter(Objects.requireNonNull<FragmentActivity>(activity), R.id.postListView, it1) }

            builderSingle.setNegativeButton("cancel") { dialog, _ -> dialog.dismiss() }

            builderSingle.setAdapter(arrayAdapter) { _, which ->
                val strName = arrayAdapter?.getItem(which)
                val builderInner = AlertDialog.Builder(activity)
                builderInner.setMessage(strName?.Title ?: "error")
                builderInner.setTitle("Your Selected Item to delete is")
                builderInner.setPositiveButton("Ok") { dialog1, _ ->
                    MainActivity.urls!!.removeAt(which)
                    (Objects.requireNonNull<FragmentActivity>(activity) as MainActivity).saveListInPreferences()
                    arrayAdapter?.notifyDataSetChanged()
                    dialog1.dismiss()
                }
                builderInner.setNegativeButton("Cancel") { dialog1, _ -> dialog1.dismiss() }
                builderInner.show()
            }
            builderSingle.show()
            true
        }
    }


}
